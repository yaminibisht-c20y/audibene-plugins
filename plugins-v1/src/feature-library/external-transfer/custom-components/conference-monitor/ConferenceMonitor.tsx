/* eslint-disable no-console */
import * as React from 'react';

import ConferenceService from '../../../../utils/serverless/Conference/ConferenceService';

class ConferenceMonitor extends React.Component<any, any> {
  state = {
    liveParticipantCount: 0,
  };

  componentDidUpdate() {
    const { task } = this.props;
    const conference = task && (task.conference || {});
    const { conferenceSid, liveParticipantCount, liveWorkerCount, participants = [] } = conference;
    const liveParticipants = participants.filter((p: any) => p.status === 'joined');

    if (liveParticipantCount > 2 && this.state.liveParticipantCount <= 2) {
      if (this.shouldUpdateParticipants(participants, liveWorkerCount)) {
        this.handleMoreThanTwoParticipants(conferenceSid, liveParticipants);
      }
      // eslint-disable-next-line sonarjs/no-collapsible-if
    } else if (liveParticipantCount <= 2 && this.state.liveParticipantCount > 2) {
      if (this.shouldUpdateParticipants(participants, liveWorkerCount)) {
        this.handleOnlyTwoParticipants(conferenceSid, liveParticipants);
      }
    }

    if (liveParticipantCount !== this.state.liveParticipantCount) {
      this.setState({ liveParticipantCount });
    }
  }

  hasUnknownParticipant = (participants = []) => {
    return participants.some((p: any) => p.participantType === 'unknown');
  };

  shouldUpdateParticipants = (participants: any, liveWorkerCount: any) => {
    console.debug(
      'dialpad-addon, ConferenceMonitor, shouldUpdateParticipants:',
      liveWorkerCount <= 1 && this.hasUnknownParticipant(participants),
    );
    return liveWorkerCount <= 1 && this.hasUnknownParticipant(participants);
  };

  handleMoreThanTwoParticipants = (conferenceSid: any, participants: any) => {
    console.log('More than two conference participants. Setting endConferenceOnExit to false for all participants.');
    this.setEndConferenceOnExit(conferenceSid, participants, false);
  };

  handleOnlyTwoParticipants = (conferenceSid: any, participants: any) => {
    console.log('Conference participants dropped to two. Setting endConferenceOnExit to true for all participants.');
    this.setEndConferenceOnExit(conferenceSid, participants, true);
  };

  setEndConferenceOnExit = async (conferenceSid: any, participants: any, endConferenceOnExit: any) => {
    const promises: any = [];
    participants.forEach((p: any) => {
      console.log(`setting endConferenceOnExit = ${endConferenceOnExit} for callSid: ${p.callSid} status: ${p.status}`);
      if (p.connecting) {
        return;
      } // skip setting end conference on connecting parties as it will fail
      promises.push(ConferenceService.setEndConferenceOnExit(conferenceSid, p.callSid, endConferenceOnExit));
    });

    try {
      await Promise.all(promises);
      console.log(`endConferenceOnExit set to ${endConferenceOnExit} for all participants`);
    } catch (error) {
      console.error(`Error setting endConferenceOnExit to ${endConferenceOnExit} for all participants\r\n`, error);
    }
  };

  render() {
    // This is a Renderless Component, only used for monitoring and taking action on conferences
    return null;
  }
}

export default ConferenceMonitor;
