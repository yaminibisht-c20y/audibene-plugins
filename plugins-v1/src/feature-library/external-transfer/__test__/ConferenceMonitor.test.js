import React from 'react';
import { shallow } from 'enzyme';
import ConferenceMonitor from "../custom-components/conference-monitor";

jest.mock("@twilio/flex-ui",()=>({
    TaskHelper: {
        isLiveCall: jest.fn(()=>{return Promise.resolve(true)})
    },
    withTheme: (M) =>{ return M},
    Manager: {
        getInstance: jest.fn(()=>{return Promise.resolve({})})
    },
    withTaskContext: (M)=>{return M}
}))

describe("test ConferenceMonitor component", () => {
    let props = {
        task:{
            conference:{
                participants:[{status: "joined"}]
            }
        },
        theme:{
            CallCanvas:{
                Button: false
            }
        },
        isOpen: true
    }
    it("component should render successfully", () => {
        const wrapper = shallow(<ConferenceMonitor {...props}/>);
        expect(wrapper).toMatchSnapshot();
    })

    let props2 = {
        task:{
        },
        theme:{
            CallCanvas:{
                Button: false
            }
        },
        isOpen: true
    }

    it("test functions", () => {
        const wrapper = shallow(<ConferenceMonitor {...props} />);
        wrapper.setState({liveParticipantCount: 2})
        wrapper.update();
        let instancew = wrapper.instance();
        instancew.componentDidUpdate();
        instancew.hasUnknownParticipant();
        instancew.shouldUpdateParticipants([{participantType:"unknown"}],1);
        instancew.handleMoreThanTwoParticipants('CF00000000000000000000000000000000', []);
        instancew.handleOnlyTwoParticipants('CF00000000000000000000000000000000', []);
        instancew.setEndConferenceOnExit('',[{connecting: true},{connecting:false}],false);
        const wrapper2 = shallow(<ConferenceMonitor {...props2} />);
        const instancew2 = wrapper2.instance();
        instancew2.componentDidUpdate();
    });
})