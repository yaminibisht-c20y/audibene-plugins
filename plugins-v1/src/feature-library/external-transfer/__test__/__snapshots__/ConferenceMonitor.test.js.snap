// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ConferenceMonitor component component should render successfully 1`] = `ShallowWrapper {}`;
